src/thunderbolt/tbtctl.c

```diff
@@ -0,0 +1,615 @@
+                return EXIT_FAILURE;
+        }
+
+        if (auth.level == AUTH_MISSING)
+                auth.level = sl;
+        if (auth.level == AUTH_KEY && auth.key == NULL)
+                auth_generate_key_string(&auth);
+
+        r = tb_device_authorize(device, &auth);
+        if (r < 0) {
+                log_error_errno(r, "Failed to authorize device: %m");
+                return EXIT_FAILURE;
+        }
+
+        if (auth.store != STORE_NONE)
+                return EXIT_SUCCESS;
+
+        r = store_put_device(store, device, &auth);
+        if (r < 0) {
+                log_error_errno(r, "Failed to commit device to store: %m");
+                return EXIT_FAILURE;
+        }
+
+        return EXIT_SUCCESS;
+}
+
+static const struct CtlCmd cmd_authorize = {
+        .name = "authorize",
+        .func = authorize_user,
+        .desc = "Authorize a thunderbolt device",
+        .root = true,
+};
+
+static int authorize_udev(struct udev *udev, int argc, char *argv[]) {
+        _cleanup_tb_device_free_ TbDevice *device = NULL;
+        _cleanup_tb_store_free_ TbStore *store = NULL;
+        _cleanup_auth_reset_ Auth auth = AUTH_INITIALIZER;
+        SecurityLevel sl;
+        int r;
+
+        if (argc < 2) {
+                fprintf(stderr, "%s: need sysfs path\n",
+                        program_invocation_short_name);
+                return EXIT_FAILURE;
+        }
+
+        r = tb_store_new(&store);
+        if (r < 0) {
+                log_error_errno(r, "Couldn't open store: %m");
+                return EXIT_FAILURE;
+        }
+
+        r = tb_device_new_from_syspath(udev, argv[1], &device);
+        if (r < 0) {
+                log_error_errno(r, "Couldn't open device: %m");
+                return EXIT_FAILURE;
+        }
+
+        r = store_get_auth(store, device->uuid, &auth);
+        if (r < 0) {
+                log_error_errno(r, "Failed to load authorization: %m");
+                return EXIT_FAILURE;
+        }
+
+        if (!auth_level_can_authorize(auth.level)) {
+                log_debug("Unknown or ignored device: %s", device->uuid);
+                /* Unknown or ignored device */
+                return EXIT_SUCCESS;
+        }
+
+        sl = tb_device_get_security_level(device);
+        if (sl < 0) {
+                log_error_errno(sl, "Failed to determine security level");
+                return EXIT_FAILURE;
+        }
+
+        if (sl != SECURITY_USER && sl != SECURITY_SECURE) {
+                log_debug("Security level of controller insufficient");
+                /* not an error here */
+                return EXIT_SUCCESS;
+        }
+
+        auth.level = MIN(auth.level, sl);
+
+        r = tb_device_authorize(device, &auth);
+        if (r < 0) {
+                log_error_errno(r, "Failed to authorize device: %m");
+                return EXIT_FAILURE;
+        }
+
+        return EXIT_SUCCESS;
+}
+
+static const struct CtlCmd cmd_udev = {
```

Comment:
The whole setup with CtlCmd seems very heavyweight... And then you have to loop over it. Why not just use the standard
```C
if (streq(cmd, "a"))
           do_a();
else if (streq(cmd, "b"))
           do_b();
else
          ...;
```

Reply:
there's also verbs.[ch]